### PR TITLE
45/Add ui and logic for editing tags

### PR DESCRIPTION
**Pull request by friedger**
This adds a UI and the logic to update tags.

Some kind of security is still needed before closing #45

### CODE REVIEW

File: app/views/partials/admin_tag_metadata.html [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<div class="container">
```
Comment (tasomaniac) [inline app/views/partials/admin_tag_metadata.html]: Should this be called `metadata`?

Comment (tasomaniac) [inline app/views/partials/admin_tag_metadata.html]: Oh I see. It is not a metadata file. It is the template for tag metadata. :)

Comment (friedger) [inline app/views/partials/admin_tag_metadata.html]: maybe we should just name the file admin_tag.html

[image: --]
Friedger Müffke
[image: http://]about.me/friedger
http://about.me/friedger

2015-09-14 12:45 GMT+02:00 Said Tahsin Dane notifications@github.com:

**Comment by tasomaniac** [comment]
Does #45 only include tag colors and descriptions? Or does it have other static info?

**Comment by friedger** [comment]
tag only has colors, title and description in the model.

[image: --]
Friedger Müffke
[image: http://]about.me/friedger
http://about.me/friedger

2015-09-14 12:51 GMT+02:00 Said Tahsin Dane notifications@github.com:

**Comment by tasomaniac** [comment]
@friedger no I meant tag colors and descriptions are the only ones missing right? What else do we need as a static data?

**Comment by Splaktar** [comment]
That should be it other than setting admin accounts in the DB and then verifying them before presenting a link to this admin panel.

**Comment by friedger** [comment]
Security has been added to the api.

For the UI, we don't have a login button anyway. Hence all admin calls are for the knowing user. We can open a new issue for the login UI.

**Comment by Splaktar** [comment]
There is a login button in the Hub UI. It uses Google+ sign in. The button doesn't display if you are logged in. We should probably have a profile image and logout option if logged in. We should open an issue for that.

**Comment by friedger** [comment]
@Splaktar Do you want to merge this?

**Comment by Splaktar** [comment]
I will take a look ASAP. I have not yet had time to do so and will not have time tonight.

File: lib/controllers/api/v1/tags.js [hunk 2]
```diff
@@ -7,9 +7,32 @@ module.exports = function (app, cacher) {
     summary: 'Returns a list of all Event Tags in the Hub Database'
   }, cacher.cache('hours', 24), utils.getModel('Tag', {}));
 
-  app.route('get', '/tags/:tagId', {
-    summary: 'Returns information on a single tag'
-  }, cacher.cache('hours', 2), utils.getModel('Tag', {
-    _id: 'tagId'
-  }, null, true));
+	app.route("get", "/tags/:tagId", {
```
Comment (Splaktar) [inline lib/controllers/api/v1/tags.js]: Please enable JSHint in your IDE as it should have flagged these double quotes.

File: lib/controllers/api/v1/tags.js [hunk 3]
```diff
@@ -7,9 +7,32 @@ module.exports = function (app, cacher) {
...
+			summary: "Returns information on a single tag"
+		}, cacher.cache('hours', 2), utils.getModel('Tag', {
+			_id: "tagId"
+		}, null, true));
+
+	app.route("put", "/tags/:tagId", {
+		  summary: "Updates a single tag, requires admin role"
+	}, middleware.auth({roles: ['admin']}), function(req, resp) {
```
Comment (Splaktar) [inline lib/controllers/api/v1/tags.js]: I get constant exceptions from this and the server won't start:

```
/Users/splaktar/Git/hub/lib/controllers/api/v1/tags.js:18
    }, middleware.auth({roles: ['admin']}), function(req, resp) {
       ^

ReferenceError: middleware is not defined
```

**Comment by Splaktar** [comment]
OK, the server starts now. I need to do additional testing to make sure that the admin check is actually effective.

**Comment by Splaktar** [comment]
I'm still working on this in the https://github.com/gdg-x/hub/tree/refactorUI branch. Made a lot of progress this weekend, but still need to do some more signin debugging.

**Comment by Splaktar** [comment]
Merged refactorUI branch into this one.
Fixed tests/CI.
Updated some libs.

Need to still do some more testing before this is ready to go though.

**Comment by Splaktar** [comment]
Seeing the following when saving an update to a tag:

```
PUT /api/v1/tags/angularjs 403 36.454 ms - 1051
ForbiddenError: invalid csrf token
    at verifytoken (/Users/splaktar/Git/gdg-x/hub/node_modules/csurf/index.js:269:11)
    at csrf (/Users/splaktar/Git/gdg-x/hub/node_modules/csurf/index.js:97:7)
    at /Users/splaktar/Git/gdg-x/hub/lib/middleware.js:59:37
    at callbacks (/Users/splaktar/Git/gdg-x/hub/node_modules/express/lib/router/index.js:164:37)
    at param (/Users/splaktar/Git/gdg-x/hub/node_modules/express/lib/router/index.js:138:11)
    at param (/Users/splaktar/Git/gdg-x/hub/node_modules/express/lib/router/index.js:135:11)
    at pass (/Users/splaktar/Git/gdg-x/hub/node_modules/express/lib/router/index.js:145:5)
    at Router._dispatch (/Users/splaktar/Git/gdg-x/hub/node_modules/express/lib/router/index.js:173:5)
    at Object.router (/Users/splaktar/Git/gdg-x/hub/node_modules/express/lib/router/index.js:33:10)
    at next (/Users/splaktar/Git/gdg-x/hub/node_modules/connect/lib/proto.js:174:15)
```

This appears to be correct since I'm not logged in. But the login button isn't appearing for me. Looking into that...

**Comment by Splaktar** [comment]
Fixed CSRF errors due to previous misconfiguration of Express and Angular (from years ago).

screen shot 2017-03-11 at 7 02 32 pm
screen shot 2017-03-11 at 7 08 04 pm


**Unfortunately, checks like `middleware.auth({roles: ['admin']})` are still not handled properly by the server. Even when I am logged out and cache/cookies cleared, I can modify the tag and its values in the DB.**

**Comment by Splaktar** [comment]
Going to open an issue for handling auth roles and not try to solve that in this PR.

**Pull request merged**